File: frontend/src/types/book.ts
/** Book genre options (synced with BE Genre enum) */
export type BookGenre = "essay" | "novel" | "poem" | "autobiography" | "children" | "non_fiction" | "other";

/** Book status in the writing pipeline (synced with BE BookStatus enum) */
export type BookStatus =
  | "draft"
  | "writing"
  | "editing"
  | "designing"
  | "completed"
  | "published";

/** Chapter editing status (synced with BE ChapterStatus enum) */
export type ChapterStatus = "draft" | "writing" | "completed" | "editing" | "finalized";

/** Main book model (matches BE BookResponse — snake_case keys) */
export interface Book {
  id: string;
  user_id: string;
  title: string;
  genre: BookGenre;
  status: BookStatus;
  description: string;
  target_audience: string;
  chapter_count: number;
  word_count: number;
  created_at: string;
  updated_at: string;
}

/** Chapter within a book (matches BE ChapterResponse — snake_case keys) */
export interface Chapter {
  id: string;
  book_id: string;
  title: string;
  content: string;
  order: number;
  status: ChapterStatus;
  word_count: number;
  created_at: string;
  updated_at: string;
}

/** Create book request */
export interface CreateBookData {
  title: string;
  genre: BookGenre;
  description?: string;
}

/** Update book request */
export interface UpdateBookData {
  title?: string;
  genre?: BookGenre;
  description?: string;
  status?: BookStatus;
}

/** Create chapter request */
export interface CreateChapterData {
  title: string;
  order: number;
}

/** Update chapter request */
export interface UpdateChapterData {
  title?: string;
  content?: string;
  order?: number;
  status?: ChapterStatus;
}

/** Editing suggestion from AI */
export interface EditSuggestion {
  id: string;
  type: "grammar" | "style" | "structure" | "content";
  original: string;
  suggested: string;
  explanation: string;
  position: {
    start: number;
    end: number;
  };
  accepted: boolean | null;
}

/** Edit history entry (FE view model — 편집 기록) */
export interface EditHistory {
  id: string;
  chapter_id: string;
  stage: "structure" | "content" | "proofread" | "final";
  suggestions: EditSuggestion[];
  applied_at: string | null;
  created_at: string;
}

/** Stage result from BE QualityReport */
export interface StageResult {
  stage: "structure" | "content" | "proofread" | "final";
  score: number;
  issues_count: number;
  feedback: string;
}

/** Quality report (matches BE QualityReport — snake_case keys) */
export interface QualityReport {
  book_id: string;
  overall_score: number;
  stage_results: StageResult[];
  total_issues: number;
  summary: string;
  recommendations: string[];
  created_at: string;
}

/** Cover design template (matches BE CoverTemplate — snake_case keys) */
export interface CoverTemplate {
  id: string;
  name: string;
  genre: BookGenre;
  style: string;
  preview_url: string;
  description: string;
}

/** Export format options */
export type ExportFormat = "docx" | "pdf" | "epub";

/** Export status (matches BE ExportStatus — snake_case keys) */
export interface ExportStatus {
  export_id: string;
  book_id: string;
  format: ExportFormat;
  status: "pending" | "processing" | "completed" | "failed";
  progress: number;
  error_message: string | null;
  created_at: string;
}

/** Export response (matches BE ExportResponse — snake_case keys) */
export interface ExportResponse {
  export_id: string;
  book_id: string;
  format: ExportFormat;
  status: "pending" | "processing" | "completed" | "failed";
  download_url: string | null;
  file_size_bytes: string | null;
  created_at: string;
}

File: backend/app/schemas/book.py
"""
도서 관련 Pydantic 스키마
도서 생성, 수정, 응답 스키마를 정의합니다.
"""

from enum import Enum

from pydantic import Field, StrictBool, StrictInt, StrictStr

from app.schemas.base import StrictBaseModel


class Genre(str, Enum):
    """장르 열거형"""
    ESSAY = "essay"                   # 수필/에세이
    NOVEL = "novel"                   # 소설
    POEM = "poem"                     # 시
    AUTOBIOGRAPHY = "autobiography"   # 자서전
    CHILDREN = "children"             # 동화
    NON_FICTION = "non_fiction"        # 논픽션
    OTHER = "other"                   # 기타


class BookStatus(str, Enum):
    """도서 상태 열거형"""
    DRAFT = "draft"           # 초안 작성 중
    WRITING = "writing"       # 집필 중
    EDITING = "editing"       # 편집 중
    DESIGNING = "designing"   # 디자인 중
    COMPLETED = "completed"   # 완성됨
    PUBLISHED = "published"   # 출판됨


class BookCreate(StrictBaseModel):
    """도서 생성 요청 스키마"""
    title: StrictStr = Field(..., min_length=1, max_length=200)
    genre: Genre
    description: StrictStr = Field(default="", max_length=2000)
    target_audience: StrictStr = Field(default="", max_length=200)


class BookUpdate(StrictBaseModel):
    """도서 수정 요청 스키마"""
    title: StrictStr | None = Field(default=None, min_length=1, max_length=200)
    genre: Genre | None = None
    description: StrictStr | None = Field(default=None, max_length=2000)
    target_audience: StrictStr | None = Field(default=None, max_length=200)
    status: BookStatus | None = None


class BookResponse(StrictBaseModel):
    """도서 응답 스키마"""
    id: StrictStr
    user_id: StrictStr
    title: StrictStr
    genre: Genre
    description: StrictStr
    target_audience: StrictStr
    status: BookStatus
    chapter_count: StrictInt
    word_count: StrictInt
    created_at: StrictStr
    updated_at: StrictStr


class BookListResponse(StrictBaseModel):
    """도서 목록 응답 스키마 (페이지네이션 포함)"""
    success: StrictBool = True
    data: list[BookResponse]
    total: StrictInt
    page: StrictInt
    page_size: StrictInt
    total_pages: StrictInt
